### ISSUE TITLE
cache-control header needed to avoid accidental caching

### ISSUE THREAD

**Issue by lutzky** [esphome/issues#2952]
### The problem

TL;DR - expected behavior:

1. Setup esphome via cloudflare
2. Install by downloading or using "plug into this computer"
3. Change something about the config
4. Install again, within 4 hours of step 2
5. Expected - behavior changes; actual - no change, because of surprise caching by cloudflare.

Details:

To make use of esphome's "Plug Into This Computer" mode, an https connection is necessary; one way of doing this is by using a proxy which would wrap the connection in https; I happened to do this using [CloudFlare Tunnels](https://developers.cloudflare.com/cloudflare-one/tutorials/single-command), which I use anyway. However, I found out that no matter how I change a config in esphome, installing it via the browser (or downloading the file and flashing it) had no effect. After a lot of head-scratching, it turned out that the files being downloaded were identical each time, despite my changes. This only happened when using cloudflare - going directly to the esphome http port worked correctly for downloading (but the "plug into this computer" mode is unsupported).

Sniffing the HTTP requests revealed the issue: The GET requests for `/download.bin` had response headers of:

* `cache-control: max-age=14400`
* `cf-cache-status: HIT`
* `content-disposition: attachment; filename="my_esphome_project.bin"

Only the `content-disposition` header is sent by esphome; the others are added by cloudflare. [It turns out](https://developers.cloudflare.com/cache/about/default-cache-behavior) that if `cache-control` isn't explicitly disabled, and the filetype has one of a few select extensions (...including `.bin`), then cloudflare happily caches it and will continue to serve the cached copy for 4 hours.

Although I'm experiencing the problem specifically with cloudflare (and I seem to be [not alone](https://community.home-assistant.io/t/setup-https-for-the-lastest-esphome-dashboard-docker-container/317415/10)), it's entirely possible that other https-adding tunnels exhibit similar behaviors, and it's worth signalling to every would-be cacher in the request path that this file may have changed recently and shouldn't be cached.

### Which version of ESPHome has the issue?

2021.12.3

### What type of installation are you using?

Docker

### What platform are you using?

ESP8266
